sql/presentation.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"crypto/md5"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// presentation constructs a presentationNode.
+func (*planner) presentation(n *parser.Select, p planNode, s *scanNode) (planNode, error) {
+	if n.Limit == nil && n.Distinct == nil {
+		return p, nil
+	}
+	// Mark sort node as not presenting the results.
+	sNode, ok := p.(*sortNode)
+	if ok {
+		sNode.dontPresent = true
+	}
+	var count, offset int64
+	if n.Limit != nil {
+		data := []struct {
+			name       string
+			src        parser.Expr
+			dst        *int64
+			defaultVal int64
+		}{
+			{"LIMIT", n.Limit.Count, &count, math.MaxInt64},
+			{"OFFSET", n.Limit.Offset, &offset, 0},
+		}
+
+		for _, datum := range data {
+			if datum.src == nil {
+				*datum.dst = datum.defaultVal
+			} else {
+				if parser.ContainsVars(datum.src) {
+					return nil, fmt.Errorf("argument of %s must not contain variables", datum.name)
+				}
+
+				normalized, err := parser.NormalizeExpr(datum.src)
+				if err != nil {
+					return nil, err
+				}
+				dstDatum, err := parser.EvalExpr(normalized)
+				if err != nil {
+					return nil, err
+				}
+
+				if dstDatum == parser.DNull {
+					*datum.dst = datum.defaultVal
+					continue
+				}
+
+				if dstDInt, ok := dstDatum.(parser.DInt); ok {
+					*datum.dst = int64(dstDInt)
+					continue
+				}
+
+				return nil, fmt.Errorf("argument of %s must be type %s, not type %s", datum.name, parser.DummyInt.Type(), dstDatum.Type())
+			}
+		}
+	}
+	pNode := &presentationNode{planNode: p, columns: p.Columns(), limit: (n.Limit != nil), count: count, offset: offset}
+
+	if n.Distinct != nil {
+		pNode.checksums = make(map[[md5.Size]byte]struct{})
+		pNode.distinctIdx = make([]int, 0, len(n.Distinct))
+		for _, expr := range n.Distinct {
+			index, err := s.maybeAddRender(expr)
+			if err != nil {
+				return nil, err
+			}
+			pNode.distinctIdx = append(pNode.distinctIdx, index)
+		}
+	}
+	return pNode, nil
+}
+
+type presentationNode struct {
+	planNode
+	// This map is surely going to overflow memory.
+	// TODO(vivek): Return an error once map has grown above a configured threshold.
```

Comment:
Rather than return an error, I think we'll want to dump the map of already seen distinct values to disk. We need to do something similar for sorting (i.e. fall back to a disk-based sort when the number of values is too large). My current thinking is that we should use local/temporary RocksDB storage for this purpose.